File: backend/app/api/v2/android_tests.py
"""
Android 测试管理 API

提供 Android 测试的 CRUD 操作接口
"""

from typing import Optional

from fastapi import APIRouter, Query
from pydantic import BaseModel

from app.api.deps import (
    PaginationDep,
)
from app.schemas.common import SuccessResponse
from app.services.android_test_service import AndroidTestService
from app.config.database import get_db
from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession


# ============ 请求模型 ============

class CreateAndroidFunctionRequest(BaseModel):
    """创建 Android 功能请求"""
    display_name: str
    name: str
    folder_id: Optional[str] = None
    description: Optional[str] = None
    app_package: Optional[str] = None
    app_activity: Optional[str] = None
    device_udid: Optional[str] = None
    business_module: Optional[str] = None
    script_format: str = "midscene"
    script_language: str = "typescript"
    test_config: Optional[dict] = None
    pages: Optional[list] = None
    tags: Optional[list] = None
    custom_config: Optional[dict] = None
    sort_order: int = 0


class UpdateAndroidFunctionRequest(BaseModel):
    """更新 Android 功能请求"""
    display_name: Optional[str] = None
    name: Optional[str] = None
    description: Optional[str] = None
    app_package: Optional[str] = None
    app_activity: Optional[str] = None
    device_udid: Optional[str] = None
    business_module: Optional[str] = None
    script_format: Optional[str] = None
    script_language: Optional[str] = None
    test_config: Optional[dict] = None
    pages: Optional[list] = None
    tags: Optional[list] = None
    custom_config: Optional[dict] = None
    sort_order: Optional[int] = None
# noqa  MC80OmFIVnBZMlhsaUpqbWxvYzZXbkpLY0E9PTo0Nzc0Zjc5MA==


# ============ 依赖注入 ============

async def get_android_test_service(
    db: AsyncSession = Depends(get_db),
) -> AndroidTestService:
    """获取 Android 测试服务实例"""
    return AndroidTestService(db)


AndroidTestServiceDep = AndroidTestService


# ============ 路由定义 ============

router = APIRouter(prefix="/projects/{project_identifier}/android-functions")


# ============ Android 功能管理接口 ============

@router.post(
    "",
    response_model=SuccessResponse,
    summary="创建 Android 功能",
    description="创建新的 Android 功能定义",
)
async def create_android_function(
    project_identifier: str,
    request: CreateAndroidFunctionRequest,
    service: AndroidTestService = Depends(get_android_test_service),
):
    """创建 Android 功能"""
    result = await service.create_android_function(
        project_identifier=project_identifier,
        **request.model_dump(),
    )

    return SuccessResponse(data=result, message="Android 功能创建成功")

# pragma: no cover  MS80OmFIVnBZMlhsaUpqbWxvYzZXbkpLY0E9PTo0Nzc0Zjc5MA==

@router.get(
    "",
    response_model=SuccessResponse,
    summary="获取 Android 功能列表",
    description="获取项目下的所有 Android 功能列表，支持搜索和过滤",
)
async def list_android_functions(
    project_identifier: str,
    pagination: PaginationDep,
    service: AndroidTestService = Depends(get_android_test_service),
    folder_id: Optional[str] = Query(None, description="文件夹 ID 过滤"),
    search: Optional[str] = Query(None, description="搜索关键词"),
):
    """获取 Android 功能列表"""
    result = await service.list_android_functions(
        project_identifier=project_identifier,
        folder_id=folder_id,
        offset=(pagination.p - 1) * pagination.page_size,
        limit=pagination.page_size,
        search=search,
    )

    return SuccessResponse(data=result)


# ============ Android 子功能管理接口 ============
# 注意：静态路由必须定义在动态路由 /{function_id} 之前，否则 FastAPI 会错误匹配

@router.get(
    "/sub-functions",
    response_model=SuccessResponse,
    summary="获取 Android 子功能列表",
    description="获取项目下的所有 Android 子功能列表，支持搜索和过滤",
)
async def list_android_sub_functions(
    project_identifier: str,
    pagination: PaginationDep,
    service: AndroidTestService = Depends(get_android_test_service),
    function_id: Optional[str] = Query(None, description="功能 ID 过滤"),
    folder_id: Optional[str] = Query(None, description="文件夹 ID 过滤"),
    search: Optional[str] = Query(None, description="搜索关键词"),
):
    """获取 Android 子功能列表"""
    result = await service.list_android_sub_functions(
        project_identifier=project_identifier,
        function_id=function_id,
        folder_id=folder_id,
        offset=(pagination.p - 1) * pagination.page_size,
        limit=pagination.page_size,
        search=search,
    )

    return SuccessResponse(data=result)


@router.get(
    "/sub-functions/{sub_function_id}",
    response_model=SuccessResponse,
    summary="获取 Android 子功能详情",
    description="获取指定 Android 子功能的详细信息",
)
async def get_android_sub_function(
    sub_function_id: str,
    service: AndroidTestService = Depends(get_android_test_service),
):
    """获取 Android 子功能详情"""
    result = await service.get_android_sub_function(sub_function_id=sub_function_id)
    return SuccessResponse(data=result)


@router.get(
    "/{function_id}",
    response_model=SuccessResponse,
    summary="获取 Android 功能详情",
    description="获取指定 Android 功能的详细信息",
)
async def get_android_function(
    function_id: str,
    service: AndroidTestService = Depends(get_android_test_service),
):
    """获取 Android 功能详情"""
    result = await service.get_android_function(function_id=function_id)
    return SuccessResponse(data=result)
# noqa  Mi80OmFIVnBZMlhsaUpqbWxvYzZXbkpLY0E9PTo0Nzc0Zjc5MA==


@router.patch(
    "/{function_id}",
    response_model=SuccessResponse,
    summary="更新 Android 功能",
    description="更新指定 Android 功能的信息",
)
async def update_android_function(
    function_id: str,
    request: UpdateAndroidFunctionRequest,
    service: AndroidTestService = Depends(get_android_test_service),
):
    """更新 Android 功能"""
    result = await service.update_android_function(
        function_id=function_id,
        **request.model_dump(exclude_none=True),
    )

    return SuccessResponse(data=result, message="Android 功能更新成功")

# pragma: no cover  My80OmFIVnBZMlhsaUpqbWxvYzZXbkpLY0E9PTo0Nzc0Zjc5MA==

@router.delete(
    "/{function_id}",
    response_model=SuccessResponse,
    summary="删除 Android 功能",
    description="删除指定的 Android 功能",
)
async def delete_android_function(
    function_id: str,
    service: AndroidTestService = Depends(get_android_test_service),
):
    """删除 Android 功能"""
    result = await service.delete_android_function(function_id=function_id)
    return SuccessResponse(data=result)


# ============ 设备管理接口 ============

devices_router = APIRouter(prefix="/projects/{project_identifier}/android-devices")


@devices_router.get(
    "",
    response_model=SuccessResponse,
    summary="获取 Android 设备列表",
    description="获取已连接的 Android 设备列表",
)
async def list_android_devices(
    project_identifier: str,
    service: AndroidTestService = Depends(get_android_test_service),
):
    """获取 Android 设备列表"""
    result = await service.list_android_devices(project_identifier=project_identifier)
    return SuccessResponse(data=result)


@devices_router.post(
    "/refresh",
    response_model=SuccessResponse,
    summary="刷新 Android 设备列表",
    description="重新扫描并获取已连接的 Android 设备列表",
)
async def refresh_android_devices(
    project_identifier: str,
    service: AndroidTestService = Depends(get_android_test_service),
):
    """刷新 Android 设备列表"""
    result = await service.refresh_android_devices(project_identifier=project_identifier)
    return SuccessResponse(data=result)

File: ui/lib/api/android-tests.ts
/**
 * Android 测试相关的 API 客户端函数
 */

import { apiClient } from "./client";

// ==================== 类型定义 ====================

export interface AndroidFunction {
  id: string;
  identifier: string;
  display_name: string;
  name: string;
  description: string | null;
  app_package: string | null;
  app_activity: string | null;
  device_udid: string | null;
  business_module: string | null;
  script_format: string;
  script_language: string;
  test_config: any;
  pages: any;
  tags: any;
  custom_config: any;
  folder_id: string | null;
  total_sub_functions: number;
  total_test_cases: number;
  total_test_runs: number;
  last_run_status: string | null;
  sort_order: number;
  created_at: string;
  updated_at: string | null;
}

export interface AndroidSubFunction {
  id: string;
  identifier: string;
  function_id: string;
  display_name: string;
  name: string;
  description: string | null;
  test_type: string;
  target_pages: any;
  test_scenario: string | null;
  test_data: any;
  expected_results: any;
  priority: string;
  tags: any;
  custom_config: any;
  folder_id: string | null;
  total_test_cases: number;
  total_test_runs: number;
  last_run_status: string | null;
  sort_order: number;
  created_at: string;
  updated_at: string | null;
}

export interface CreateAndroidFunctionRequest {
  display_name: string;
  name: string;
  folder_id?: string;
  description?: string;
  app_package?: string;
  app_activity?: string;
  device_udid?: string;
  business_module?: string;
  script_format?: string;
  script_language?: string;
  test_config?: any;
  pages?: any;
  tags?: any;
  custom_config?: any;
  sort_order?: number;
}
// NOTE  MC80OmFIVnBZMlhsaUpqbWxvYzZTWFJxVGc9PTowYjVlM2NjOQ==

export interface UpdateAndroidFunctionRequest {
  display_name?: string;
  name?: string;
  description?: string;
  app_package?: string;
  app_activity?: string;
  device_udid?: string;
  business_module?: string;
  script_format?: string;
  script_language?: string;
  test_config?: any;
  pages?: any;
  tags?: any;
  custom_config?: any;
  sort_order?: number;
}

export interface CreateAndroidSubFunctionRequest {
  function_id: string;
  display_name: string;
  name: string;
  folder_id?: string;
  description?: string;
  test_type?: string;
  target_pages?: any;
  test_scenario?: string;
  test_data?: any;
  expected_results?: any;
  priority?: string;
  tags?: any;
  custom_config?: any;
  sort_order?: number;
}

export interface UpdateAndroidSubFunctionRequest {
  display_name?: string;
  name?: string;
  description?: string;
  test_type?: string;
  target_pages?: any;
  test_scenario?: string;
  test_data?: any;
  expected_results?: any;
  priority?: string;
  tags?: any;
  custom_config?: any;
  sort_order?: number;
}

export interface AndroidFunctionListResponse {
  items: AndroidFunction[];
  total: number;
  offset: number;
  limit: number;
}

export interface AndroidSubFunctionListResponse {
  items: AndroidSubFunction[];
  total: number;
  offset: number;
  limit: number;
}

export interface AndroidDevice {
  udid: string;
  name: string;
  model: string;
  version: string;
  status: "connected" | "disconnected" | "busy" | "offline";
  screen_resolution?: string;
  dpi?: number;
}

// ==================== Android 功能 API 函数 ====================
// @ts-expect-error  MS80OmFIVnBZMlhsaUpqbWxvYzZTWFJxVGc9PTowYjVlM2NjOQ==

/**
 * 获取项目的 Android 功能列表
 */
export async function listAndroidFunctions(
  projectIdentifier: string,
  params?: {
    p?: number;
    page_size?: number;
    folder_id?: string;
    search?: string;
  }
): Promise<AndroidFunctionListResponse> {
  const response = await apiClient.get<{ data: AndroidFunctionListResponse }>(
    `/projects/${projectIdentifier}/android-functions`,
    { params }
  );
  return response.data;
}

/**
 * 获取 Android 功能详情
 */
export async function getAndroidFunction(
  projectIdentifier: string,
  functionId: string
): Promise<AndroidFunction> {
  const response = await apiClient.get<{ data: AndroidFunction }>(
    `/projects/${projectIdentifier}/android-functions/${functionId}`
  );
  return response.data;
}

/**
 * 创建 Android 功能
 */
export async function createAndroidFunction(
  projectIdentifier: string,
  data: CreateAndroidFunctionRequest
): Promise<AndroidFunction> {
  const response = await apiClient.post<{ data: AndroidFunction }>(
    `/projects/${projectIdentifier}/android-functions`,
    data
  );
  return response.data;
}
// eslint-disable  Mi80OmFIVnBZMlhsaUpqbWxvYzZTWFJxVGc9PTowYjVlM2NjOQ==

/**
 * 更新 Android 功能
 */
export async function updateAndroidFunction(
  projectIdentifier: string,
  functionId: string,
  data: UpdateAndroidFunctionRequest
): Promise<AndroidFunction> {
  const response = await apiClient.patch<{ data: AndroidFunction }>(
    `/projects/${projectIdentifier}/android-functions/${functionId}`,
    data
  );
  return response.data;
}

/**
 * 删除 Android 功能
 */
export async function deleteAndroidFunction(
  projectIdentifier: string,
  functionId: string
): Promise<{ success: boolean; message: string }> {
  const response = await apiClient.delete<{ data: { success: boolean; message: string } }>(
    `/projects/${projectIdentifier}/android-functions/${functionId}`
  );
  return response.data;
}

// ==================== Android 子功能 API 函数 ====================

/**
 * 获取项目的 Android 子功能列表
 */
export async function listAndroidSubFunctions(
  projectIdentifier: string,
  params?: {
    p?: number;
    page_size?: number;
    function_id?: string;
    folder_id?: string;
    search?: string;
  }
): Promise<AndroidSubFunctionListResponse> {
  const response = await apiClient.get<{ data: AndroidSubFunctionListResponse }>(
    `/projects/${projectIdentifier}/android-functions/sub-functions`,
    { params }
  );
  return response.data;
}

/**
 * 获取 Android 子功能详情
 */
export async function getAndroidSubFunction(
  projectIdentifier: string,
  subFunctionId: string
): Promise<AndroidSubFunction> {
  const response = await apiClient.get<{ data: AndroidSubFunction }>(
    `/projects/${projectIdentifier}/android-functions/sub-functions/${subFunctionId}`
  );
  return response.data;
}
// FIXME  My80OmFIVnBZMlhsaUpqbWxvYzZTWFJxVGc9PTowYjVlM2NjOQ==

/**
 * 创建 Android 子功能
 */
export async function createAndroidSubFunction(
  projectIdentifier: string,
  data: CreateAndroidSubFunctionRequest
): Promise<AndroidSubFunction> {
  const response = await apiClient.post<{ data: AndroidSubFunction }>(
    `/projects/${projectIdentifier}/android-functions/sub-functions`,
    data
  );
  return response.data;
}

/**
 * 更新 Android 子功能
 */
export async function updateAndroidSubFunction(
  projectIdentifier: string,
  subFunctionId: string,
  data: UpdateAndroidSubFunctionRequest
): Promise<AndroidSubFunction> {
  const response = await apiClient.patch<{ data: AndroidSubFunction }>(
    `/projects/${projectIdentifier}/android-functions/sub-functions/${subFunctionId}`,
    data
  );
  return response.data;
}

/**
 * 删除 Android 子功能
 */
export async function deleteAndroidSubFunction(
  projectIdentifier: string,
  subFunctionId: string
): Promise<{ success: boolean; message: string }> {
  const response = await apiClient.delete<{ data: { success: boolean; message: string } }>(
    `/projects/${projectIdentifier}/android-functions/sub-functions/${subFunctionId}`
  );
  return response.data;
}

// ==================== 设备管理 API 函数 ====================

/**
 * 获取已连接的 Android 设备列表
 */
export async function listAndroidDevices(
  projectIdentifier: string
): Promise<AndroidDevice[]> {
  const response = await apiClient.get<{ data: AndroidDevice[] }>(
    `/projects/${projectIdentifier}/android-devices`
  );
  return response.data;
}

/**
 * 刷新设备列表（扫描 ADB 设备）
 */
export async function refreshAndroidDevices(
  projectIdentifier: string
): Promise<AndroidDevice[]> {
  const response = await apiClient.post<{ data: AndroidDevice[] }>(
    `/projects/${projectIdentifier}/android-devices/refresh`
  );
  return response.data;
}
